### PR TITLE
MySqlUserRepository

### LINKED ISSUE #178: MySQL User Repository
refer to https://github.com/vinyldns/vinyldns/issues/16

1. Create MySQLUserRepository that implements functionality of UserRepository
1. Create integration tests

### PR DESCRIPTION
**Pull request by nimaeskandary**
Fixes #178 .

Changes in this pull request:
- Adds MySQL implementation of user repo

### CODE REVIEW

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlDataStoreProvider.scala [hunk 1]
```diff
@@ -41,7 +41,7 @@ class MySqlDataStoreProvider extends DataStoreProvider {
       _ <- validateRepos(config.repositories)
       _ <- runDBMigrations(settingsConfig)
       _ <- setupDBConnection(settingsConfig)
-      store <- initializeRepos()
+      store <- initializeRepos(cryptoAlgebra)
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlDataStoreProvider.scala]: `private val implementedRepositories` needs to be updated (line 35)

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlUserRepository.scala [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  final val BASE_GET_USERS: String =
+    """
+      | SELECT data
+      |  FROM user
+      |  WHERE id IN
+      """.stripMargin
+
+  final def buildGetUsersQuery(set: Set[String]): String = {
+    val stringBuilder = new StringBuilder().append("(")
+    set.toList.zipWithIndex.foreach {
```
Comment (mitruly) [inline modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlUserRepository.scala]: This works, but seems a bit of work to use both `StringBuilder` and zipping. Here's an implementation just using string interpolation and `fold`:
```
val setToList = set.toList
val stringToAppend = s"${setToList.dropRight(1).fold("(")((acc, cur) => acc + s"'$cur', ")}'${setToList.last}')"
```

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlUserRepository.scala [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      case (element, count) if count < set.size - 1 => stringBuilder.append(s"'$element', ")
```
Comment (pauljamescleary) [inline modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlUserRepository.scala]: Im going to go ahead and say generating question marks is simpler...

```scala
          val inClause = " IN (" + fqdns.as("?").mkString(",") + ")"
          val query = FIND_BY_FQDN + inClause
          SQL(query).bind(fqdns.map(_.value): _*).map(toRecordSet).list().apply()
```

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlUserRepository.scala [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      startFrom: Option[String],
+      maxItems: Option[Int]): IO[ListUsersResults] =
+    monitor("repo.User.getUsers") {
+      logger.info(s"Getting users with ids: $userIds")
+      IO {
+        if (userIds.isEmpty)
+          ListUsersResults(List[User](), None)
+        else {
+          val users = DB.readOnly { implicit s =>
+            val inClause = " IN (" + userIds.toList.as("?").mkString(",") + ")"
```
Comment (pauljamescleary) [inline modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlUserRepository.scala]: I am good with this as it is consistent.

**Pull request merged**